crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -289,15 +289,40 @@ fn validate_inbounds<T: ComponentType>(memory: &[u8], ptr: &ValRaw) -> Result<us
     Ok(ptr)
 }
 
-unsafe fn handle_result(func: impl FnOnce() -> Result<()>) {
-    match crate::runtime::vm::catch_unwind_and_longjmp(func) {
+unsafe fn call_host_and_handle_result<T>(
+    cx: *mut VMOpaqueContext,
+    func: impl FnOnce(
+        *mut ComponentInstance,
+        &Arc<ComponentTypes>,
+        StoreContextMut<'_, T>,
+    ) -> Result<()>,
+) {
+    let cx = VMComponentContext::from_opaque(cx);
+    let instance = (*cx).instance();
+    let types = (*instance).component_types();
+    let raw_store = (*instance).store();
+    let store = StoreContextMut::from_raw(raw_store);
+
+    let res = store
+        .0
+        .call_hook(CallHook::CallingHost)
+        .and_then(|_| crate::runtime::vm::catch_unwind_and_longjmp(|| func(instance, types, store)))
```

Comment:
Could you expand the `catch_unwind_and_longjmp` to encompass the invocation of the call hooks too?